### PR TITLE
Implement hasher

### PR DESCRIPTION
**Pull request by justing-bq**
Provide implementation for stubbed hasher functions.

### CODE REVIEW

File: err.go [hunk 1]
```diff
@@ -21,15 +21,20 @@ import "fmt"
 type InvalidOperationError struct {
 	structName string
 	methodName string
+	message    string
 }
 
 func (e *InvalidOperationError) Error() string {
-	return fmt.Sprintf(`ionhash: invalid operation error on %v.%v`, e.structName, e.methodName)
+	if e.message != "" {
+		return fmt.Sprintf(`ionhash: %v.%v: %v`, e.structName, e.methodName, e.message)
```
Comment (therapon) [inline err.go:29]: For the generic case I would add something 
```
`ionhash: Invalid operation at %v,%v: %v`
```

To distinguish if from just printing information.

File: hasher.go [hunk 2]
```diff
@@ -30,22 +32,66 @@ func newHasher(hasherProvider IonHasherProvider) *hasher {
...
+
+	h.hasherStack.push(h.currentHasher)
+	h.currentHasher.stepIn(ionValue)
 }
 
-func (h *hasher) stepOut() {
-	panic("implement me")
+func (h *hasher) stepOut() error {
+	if h.depth() == 0 {
+		return &InvalidOperationError{"hasher", "stepOut", "Hasher cannot step out any further"}
```
Comment (therapon) [inline hasher.go:61]: I woiuld add to the message that depth is 0. That is the reason for not being able to step out and we should.

File: hasher.go [hunk 3]
```diff
@@ -30,22 +32,66 @@ func newHasher(hasherProvider IonHasherProvider) *hasher {
...
+
+	return nil
 }
 
-func (h *hasher) digest() []byte {
-	panic("implement me")
+func (h *hasher) digest() ([]byte, error) {
+	if h.depth() == 0 {
+		return nil, &InvalidOperationError{
+			"hasher", "stepOut", "A digest may only be provided at the same depth hashing started"}
```
Comment (therapon) [inline hasher.go:89]: This is not `stepOut` this is `digest`
I am not sure I get the condition either that leads to the error either. If the value is a single integer wouldn't that be at depth 0?

File: stack.go [hunk 4]
```diff
@@ -27,14 +27,30 @@ func (s *stack) push(element interface{}) {
 	*s = append(*s, element)
 }
 
-// Remove and return top element of stack. Return false if stack is empty.
+// Remove and return top element of stack. Return error if stack is empty.
 func (s *stack) pop() (interface{}, error) {
 	if s.isEmpty() {
-		return nil, &InvalidOperationError{"stack", "Pop"}
+		return nil, &InvalidOperationError{structName: "stack", methodName: "pop"}
```
Comment (therapon) [inline stack.go:33]: We should be able to pass some more information here. pop is invalid because the stack is empty.

File: stack.go [hunk 5]
```diff
@@ -27,14 +27,30 @@ func (s *stack) push(element interface{}) {
...
 		element := (*s)[index] // Index into the slice and obtain the element.
 		*s = (*s)[:index]      // Remove it from the stack by slicing it off.
 		return element, nil
 	}
 }
+
+// Return top element of stack. Return error if stack is empty.
+func (s *stack) peek() (interface{}, error) {
+	if s.isEmpty() {
+		return nil, &InvalidOperationError{structName: "stack", methodName: "peek"}
```
Comment (therapon) [inline stack.go:45]: ditto about passing more information here.

File: err.go (hunk 1 above)
Comment (justing-bq) [reply err.go]: Done.

File: hasher.go (hunk 2 above)
Comment (justing-bq) [reply hasher.go]: Done.

File: hasher.go (hunk 3 above)
Comment (justing-bq) [reply hasher.go]: Fixed stepOut to digest.
I also fixed the condition to h.depth() != 0.

File: stack.go (hunk 4 above)
Comment (justing-bq) [reply stack.go]: Done.

File: stack.go (hunk 5 above)
Comment (justing-bq) [reply stack.go]: Done.

**Pull request merged**